File: targets/node/koa/routes/cached.js
// cached: validator headers and the conditional request.
//
// The ETag is pinned in the fixture, so this measures emitting the header and comparing it
// rather than hashing the body. Koa's ctx.fresh would work from an ETag it was given, but
// it also consults Cache-Control and Last-Modified, and this family is only about the one
// comparison.
//
// The comparison requires a non-empty header: matching a missing if-none-match against an
// empty ETag answers 304 to a client that never asked a conditional question.
import * as d from "../../_shared/domain.js";

const validators = (size) => {
  const etag = d.etagOf(size);
  return async (ctx, next) => {
    ctx.set({ etag, "cache-control": d.CACHEABLE, "x-rb-serial": d.nextSerial() });
    if (ctx.headers["if-none-match"] === etag) {
      ctx.status = 304;
      return;
    }
    await next();
  };
};

export default function cached(router) {
  // rb:snippet cached.small cached.medium cached.large cached.revalidate
  for (const size of ["small", "medium", "large"]) {
    router.get("/cached/" + size, validators(size), (ctx) => { ctx.body = d.payload(size); });
  }
}

File: targets/java/quarkus/src/main/java/rb/quarkus/routes/Cached.java
package rb.quarkus.routes;

import jakarta.ws.rs.GET;
import jakarta.ws.rs.HeaderParam;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import rb.domain.Domain;

/**
 * cached: validator headers and the conditional request.
 *
 * The ETag is pinned in the fixture, so this measures emitting the header and comparing it
 * rather than hashing the body. JAX-RS's Request.evaluatePreconditions would work from an
 * EntityTag it built, and it also consults Last-Modified; this family is only about the one
 * comparison.
 *
 * The comparison requires a non-empty header: matching a missing if-none-match against an
 * empty ETag answers 304 to a client that never asked a conditional question.
 */
@Path("/cached")
@Produces(MediaType.APPLICATION_JSON)
public class Cached {

  private static Response serve(String size, String inm) {
    String etag = Domain.etagOf(size);
    Response.ResponseBuilder b = inm != null && inm.equals(etag)
        ? Response.status(304)
        : Response.ok(Domain.payload(size));
    return b.header("etag", etag)
            .header("cache-control", Domain.CACHEABLE)
            .header("x-rb-serial", Domain.nextSerial())
            .build();
  }

  // rb:snippet cached.small
  @GET
  @Path("small")
  public Response small(@HeaderParam("if-none-match") String inm) {
    return serve("small", inm);
  }

  // rb:snippet cached.medium
  @GET
  @Path("medium")
  public Response medium(@HeaderParam("if-none-match") String inm) {
    return serve("medium", inm);
  }

  // rb:snippet cached.large cached.revalidate
  @GET
  @Path("large")
  public Response large(@HeaderParam("if-none-match") String inm) {
    return serve("large", inm);
  }
}
